### PR TITLE
CFE-2882: Don't trim leading zeroes in ubuntu minor version class

### PR DESCRIPTION
**Pull request by olehermanse**
# Ubuntu 18

**Before:**
```
root@ip-172-31-45-190 core $ cf-promises --show-classes | grep -i ubuntu
linux_x86_64_4_15_0_1032_aws__34_Ubuntu_SMP_Thu_Jan_17_15_18_09_UTC_2019 source=agent,derived-from=sys.long_arch,hardclass
ubuntu                                                       inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
ubuntu_18                                                    inventory,attribute_name=none,source=agent,derived-from=sys.flavor,hardclass
ubuntu_18_04                                                 inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
ubuntu_18_4                                                  inventory,attribute_name=none,source=agent,hardclass
root@ip-172-31-45-190 core $ cf-promises --show-classes | grep -i debian
debian                                                       inventory,attribute_name=none,source=agent,hardclass
debian_buster                                                inventory,attribute_name=none,source=agent,hardclass
debian_derived                                               source=promise,inventory,attribute_name=none
root@ip-172-31-45-190 core $ 
```

**After:**
```
root@ip-172-31-45-190 core $ cf-promises --show-classes | grep -i ubuntu
linux_x86_64_4_15_0_1032_aws__34_Ubuntu_SMP_Thu_Jan_17_15_18_09_UTC_2019 source=agent,derived-from=sys.long_arch,hardclass
ubuntu                                                       inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
ubuntu_18                                                    inventory,attribute_name=none,source=agent,derived-from=sys.flavor,hardclass
ubuntu_18_04                                                 inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
root@ip-172-31-45-190 core $ cf-promises --show-classes | grep -i debian
debian                                                       inventory,attribute_name=none,source=agent,derived-from-file=/etc/debian_version,hardclass
debian_buster                                                inventory,attribute_name=none,source=agent,derived-from-file=/etc/debian_version,hardclass
debian_derived                                               source=promise,inventory,attribute_name=none
root@ip-172-31-45-190 core $ 
```

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2241/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2241/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2241/

**Comment by olehermanse** [comment]
# Ubuntu 14

**Before:**

```
root@ip-172-31-33-186:~/core# /var/cfengine/bin/cf-promises --show-classes | grep -i ubuntu
linux_x86_64_3_13_0_161_generic__211_Ubuntu_SMP_Wed_Oct_3_14_52_35_UTC_2018 source=agent,derived-from=sys.long_arch,hardclass
ubuntu                                                       inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
ubuntu_14                                                    inventory,attribute_name=none,source=agent,derived-from=sys.flavor,hardclass
ubuntu_14_04                                                 inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
ubuntu_14_4                                                  inventory,attribute_name=none,source=agent,hardclass
root@ip-172-31-33-186:~/core# /var/cfengine/bin/cf-promises --show-classes | grep -i debian
debian                                                       inventory,attribute_name=none,source=agent,hardclass
debian_derived                                               source=promise,inventory,attribute_name=none
debian_jessie                                                inventory,attribute_name=none,source=agent,hardclass
root@ip-172-31-33-186:~/core# 
```

**After:**
```
root@ip-172-31-33-186:~/core# /var/cfengine/bin/cf-promises --show-classes | grep -i ubuntu
linux_x86_64_3_13_0_161_generic__211_Ubuntu_SMP_Wed_Oct_3_14_52_35_UTC_2018 source=agent,derived-from=sys.long_arch,hardclass
ubuntu                                                       inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
ubuntu_14                                                    inventory,attribute_name=none,source=agent,derived-from=sys.flavor,hardclass
ubuntu_14_04                                                 inventory,attribute_name=none,source=agent,derived-from-file=/etc/os-release,hardclass
root@ip-172-31-33-186:~/core# /var/cfengine/bin/cf-promises --show-classes | grep -i debian
debian                                                       inventory,attribute_name=none,source=agent,derived-from-file=/etc/debian_version,hardclass
debian_derived                                               source=promise,inventory,attribute_name=none
debian_jessie                                                inventory,attribute_name=none,source=agent,derived-from-file=/etc/debian_version,hardclass
root@ip-172-31-33-186:~/core# 
```

**Comment by olehermanse** [comment]
@nickanderson @vpodzime Are you okay with these changes? It's technically a breaking change, although pretty minor (only affects you if you're using `ubuntu_18_4` or similar classes). I think we should cherry-pick to 3.12.x, but not further back. Agree?

**Comment by olehermanse** [comment]
@nickanderson Please review the acceptance test.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2242/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2242/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2242/

**Comment by nickanderson** [comment]
It looks better after, but I'm unsure of removing the classes for existing distro releases for backwards compatibility.

File: tests/acceptance/02_classes/03_os/ubuntu.cf [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+#######################################################
+
+bundle agent init
+{
+  vars:
+      "dummy" string => "dummy";
+}
+
+#######################################################
```
Comment (vpodzime) [inline tests/acceptance/02_classes/03_os/ubuntu.cf]: I think you don't need this `init` bundle. Maybe not even `test`?

**Comment by olehermanse** [comment]
> It looks better after, but I'm unsure of removing the classes for existing distro releases for backwards compatibility.

My response to this would be that we don't backport it too far. The change is very small, and easy to adjust to (just grep all your policy for `ubuntu_18_4`, `ubuntu_16_4`, `ubuntu_14_4`, `ubuntu_12_4`).

I don't think we should make a special case for Ubuntu 18+. It makes it less consistent and more confusing for those who are using or starting to use the new version. My suggestion is to change it in 3.12.x and onwards. Keep 3.10.x the same.

File: tests/acceptance/02_classes/03_os/ubuntu.cf (hunk 1 above)
Comment (olehermanse) [reply tests/acceptance/02_classes/03_os/ubuntu.cf]: I am very unsure about the "best way" to make these tests. I asked nick to review and he approved, so I guess it's okay ;)

**Comment by nickanderson** [comment]
Ok, Im on board. The commit explains it fully, and the changelog entry is set to be the commit, so the necessary information will make it to end users (if they read changelogs).

**Pull request merged**